File: public/engine/motion.ts
// The scheduler's motion input: what the reader is doing, published by the
// strip that owns the scroller once per coalesced scroll frame.
//
// The engine cannot derive any of this itself. It sees canvases appear and
// disappear and render requests arrive; it has no idea whether the reader is
// parked on a page or throwing the document at twelve thousand pixels a
// second. The virtualizer does know — it owns the scroll container, the
// geometry and the smoothed velocity — so it publishes the answer, and the
// scheduler spends its lanes on it.
//
// Everything here is state, not policy: the numbers arrive over the bridge
// (`setScrollMotion`, `configureMotion` in pdfEngine.ts) and the scheduler and
// the memory ledger read them. The one piece of interpretation this module
// owns is the phase name, because the wire spelling is a contract with the
// Rust side and belongs in exactly one place per language — the host test in
// crates/pdf-engine/tests/engine_contract.rs pins the two together.

/** The wire spellings, in ascending order of movement. `MotionPhase::wire`
 *  in crates/pdf-engine/src/api/motion.rs declares the same five. */
export const PHASES = ["idle", "slow", "normal", "fast", "fling"] as const;

/** How fast the scroller is moving, as the virtualizer classified it. */
export type Phase = (typeof PHASES)[number];

/** A 1-based inclusive page range. */
export type PageWindow = { first: number; last: number };

/** One scroll frame's motion. */
export type MotionState = {
  phase: Phase;
  /** Sign of the smoothed velocity: 1 forward through the document, -1 back,
   *  0 at rest. The magnitude itself does not cross the bridge — the phase is
   *  the speed, classified, and nothing on this side would spend the number on
   *  anything else. */
  direction: number;
  /** The page the reader is projected to reach. 0 when unpublished. */
  predictedPage: number;
  /** Pages owed a full raster. Null means "no window published". */
  full: PageWindow | null;
  /** Pages owed at least a preview raster. Null means "no window published". */
  preview: PageWindow | null;
  /** How long a newly requested render waits before it may run. */
  delayMs: number;
  /** How many raster lanes to run. 0 means "the engine's own default". */
  workers: number;
};

/** The raster budget the memory ledger enforces (public/engine/memory.ts). */
export type MotionBudget = {
  /** Every surface the engine holds, bytes. */
  maxBytes: number;
  /** The preview tier's own ceiling, bytes. */
  previewBytes: number;
  /** Hard ceiling on preview surfaces, whatever they cost. */
  maxPreviewPages: number;
  /** Output-scale multiplier a preview raster is rendered at. */
  previewScale: number;
};

/** The state a document starts in, and the one a mode that publishes no
 *  motion (single page, spread) leaves in place: nothing is being paced, no
 *  window is closed, so every request is owed a full raster at the engine's
 *  own default lane count. */
export const IDLE_MOTION: MotionState = {
  phase: "idle",
  direction: 0,
  predictedPage: 0,
  full: null,
  preview: null,
  delayMs: 0,
  workers: 0,
};

export const DEFAULT_BUDGET: MotionBudget = {
  maxBytes: 96 * 1024 * 1024,
  previewBytes: 24 * 1024 * 1024,
  maxPreviewPages: 8,
  previewScale: 0.45,
};

/** A motion publication older than this is not a description of the present.
 *  The scheduler only cancels work on the strength of a FRESH moving frame:
 *  a zoom holds scroll feedback off for the length of its tween, and a
 *  window published before it would otherwise be used to cancel the renders
 *  the zoom's own commit just asked for. */
const MOTION_TTL_MS = 400;

let motion: MotionState = IDLE_MOTION;
let budget: MotionBudget = { ...DEFAULT_BUDGET };
let generation = 0;
let publishedAt = 0;
let live = false;

/** Prediction telemetry: the last page a moving frame predicted, and how the
 *  reader's next rest compared to it. The hit rate is the number the
 *  projection horizon should be tuned against — a predictor nobody believes
 *  is just latency. */
let lastPredicted = 0;
let predictions = 0;
let predictionHits = 0;

function parsePhase(value: string): Phase {
  const found = PHASES.find((phase) => phase === value);
  return found ?? "idle";
}

/** A page range off the wire, where `0,0` spells "nothing published". */
function parseWindow(first: number, last: number): PageWindow | null {
  if (!(first > 0) || !(last >= first)) return null;
  return { first, last };
}

/** Adopt the movement half of one frame. Called from the facade's
 *  `setScrollMotion`; the tier windows arrive right behind it over
 *  `setRenderTiers`, which is what re-scores the queue. */
export function setMotion(
  phase: string,
  direction: number,
  predictedPage: number,
  delayMs: number,
  workers: number,
): void {
  const next: MotionState = {
    ...motion,
    phase: parsePhase(phase),
    direction: direction > 0 ? 1 : direction < 0 ? -1 : 0,
    predictedPage: predictedPage > 0 ? Math.floor(predictedPage) : 0,
    delayMs: delayMs > 0 ? Math.floor(delayMs) : 0,
    workers: workers > 0 ? Math.floor(workers) : 0,
  };
  notePrediction(next);
  motion = next;
  publishedAt = Date.now();
  live = true;
}

/** Adopt the geometry half of one frame: the two tier windows, 1-based and
 *  inclusive, where `0,0` means "unpublished" and leaves the tier open. */
export function setTiers(
  firstFull: number,
  lastFull: number,
  firstPreview: number,
  lastPreview: number,
): void {
  motion = {
    ...motion,
    full: parseWindow(firstFull, lastFull),
    preview: parseWindow(firstPreview, lastPreview),
  };
  generation += 1;
  publishedAt = Date.now();
  live = true;
}

/** Score one prediction against the rest that followed it. A frame that
 *  stops moving is the reader arriving, so the page they arrived at is the
 *  answer the previous frame's projection can be graded against. */
function notePrediction(next: MotionState): void {
  if (motion.phase !== "idle" && motion.predictedPage > 0) {
    lastPredicted = motion.predictedPage;
  }
  if (next.phase !== "idle" || lastPredicted === 0 || next.predictedPage === 0) return;
  predictions += 1;
  if (Math.abs(lastPredicted - next.predictedPage) <= 1) predictionHits += 1;
  lastPredicted = 0;
}

/** The current motion. Never null: an unpublished document reads as idle. */
export function currentMotion(): MotionState {
  return motion;
}

/** Bumped once per publication. A queued render records the generation it was
 *  asked for at, and the scheduler uses the gap to tell a request the reader
 *  has already outrun from one that is still current. */
export function motionGeneration(): number {
  return generation;
}

/** Whether the motion in hand may be acted on: a strip published it, and it
 *  is recent enough to still describe the reader. */
export function motionIsCurrent(): boolean {
  return live && Date.now() - publishedAt <= MOTION_TTL_MS;
}

export function currentBudget(): MotionBudget {
  return budget;
}

/** Adopt the raster budget. Called once per document from the facade's
 *  `configureMotion`; the defaults stand until then. */
export function configureBudget(
  maxBytes: number,
  previewBytes: number,
  maxPreviewPages: number,
  previewScale: number,
): void {
  budget = {
    maxBytes: maxBytes > 0 ? maxBytes : DEFAULT_BUDGET.maxBytes,
    previewBytes: previewBytes > 0 ? Math.min(previewBytes, maxBytes) : DEFAULT_BUDGET.previewBytes,
    maxPreviewPages: maxPreviewPages > 0 ? Math.floor(maxPreviewPages) : DEFAULT_BUDGET.maxPreviewPages,
    previewScale: previewScale > 0 && previewScale <= 1 ? previewScale : DEFAULT_BUDGET.previewScale,
  };
}

/** Prediction telemetry, for `stats()`. */
export function predictionStats(): { predictions: number; hits: number } {
  return { predictions, hits: predictionHits };
}

/** Forget the document's motion: called on open and on teardown, so a book
 *  that was being flung cannot pace the first renders of the next one. */
export function resetMotion(): void {
  motion = IDLE_MOTION;
  generation += 1;
  publishedAt = 0;
  live = false;
  lastPredicted = 0;
  predictions = 0;
  predictionHits = 0;
}

/** Whether `page` falls inside `window`. An unpublished window contains
 *  nothing, which is the honest answer: a caller that never said what it owes
 *  has not asked the scheduler to hold anything back — every gate that
 *  excludes work is also gated on [`motionIsCurrent`]. */
export function inWindow(page: number, window: PageWindow | null): boolean {
  return !!window && page >= window.first && page <= window.last;
}

/** Distance from `page` to the reader's heading, weighted against travel.
 *
 * A page behind the reader costs three times what the same distance ahead
 * does: reading is directional, and the page just read is only coming back if
 * the reader changes their mind. `anchor` of 0 means "unknown", and then the
 * term is dropped rather than measured from page 1. */
export function directionalDistance(page: number, anchor: number, direction: number): number {
  if (!(anchor > 0)) return 0;
  const delta = page - anchor;
  if (direction > 0) return delta >= 0 ? delta : -delta * 3;
  if (direction < 0) return delta <= 0 ? -delta : delta * 3;
  return Math.abs(delta);
}

File: src/features/reader/motion.rs
//! The reader's motion policy, and the bridge that carries it to the engine.
//!
//! One scheduler is split across the wasm boundary, because the two halves
//! know different things. The virtualizer knows the geometry and the movement:
//! how fast the reader is going, which way, which page they are projected to
//! reach, and which pages are owed what. The engine knows what a raster costs
//! and holds the only queue in front of pdf.js. Neither can prioritise a render
//! on its own, so the strip publishes the frame and the engine spends its lanes
//! on it.
//!
//! What crosses is deliberately small and quantized: a phase, a sign, five page
//! numbers, a delay and a lane count. No offsets (the engine has no geometry),
//! no raw velocity (the phase IS the speed, classified), and nothing that
//! changes on a frame where the answer did not — [`use_motion_bridge`] compares
//! the whole payload and stays quiet when a scroll moved the reader inside the
//! same page.

use leptos::prelude::*;
use virtual_list::Window;
use virtual_list_leptos::{AdaptivePolicy, RenderPlan, ScrollPhase, Virtualizer};

use pdf_engine::api as engine;
use pdf_engine::api::motion::{MotionBudget, MotionPhase, ScrollMotion};

/// The policy both scrolling strips run, and the source of every number the
/// engine is given about tiers and memory. One constant so the two sides of
/// the boundary cannot drift: the virtualizer's windows and the engine's
/// raster budget are the same policy, read twice.
pub(crate) const POLICY: AdaptivePolicy = AdaptivePolicy::reader();

/// The classified movement, spelled the way the engine parses it. The two
/// enums live in crates that do not depend on each other, and this match is
/// the one place they meet — `crates/pdf-engine/tests/engine_contract.rs` pins
/// the strings against the engine's own table.
fn wire_phase(phase: ScrollPhase) -> MotionPhase {
    match phase {
        ScrollPhase::Idle => MotionPhase::Idle,
        ScrollPhase::Slow => MotionPhase::Slow,
        ScrollPhase::Normal => MotionPhase::Normal,
        ScrollPhase::Fast => MotionPhase::Fast,
        ScrollPhase::Fling => MotionPhase::Fling,
    }
}

/// One item window as the engine reads it: 1-based inclusive page numbers.
fn wire_window(window: Option<Window>) -> Option<(u32, u32)> {
    window.map(|window| (window.first as u32 + 1, window.last as u32 + 1))
}

/// The frame's plan, translated. A plan with no mount window is a document
/// with no pages, and the honest translation of that is "no motion published"
/// rather than a prediction of page 1.
fn wire_motion(plan: RenderPlan) -> ScrollMotion {
    if plan.mount.is_none() {
        return ScrollMotion::idle();
    }
    ScrollMotion {
        phase: wire_phase(plan.phase),
        direction: plan.direction,
        predicted_page: plan.predicted_index as u32 + 1,
        full: wire_window(plan.full),
        preview: wire_window(plan.preview),
        delay_ms: plan.delay_ms,
        workers: plan.workers as u32,
    }
}

/// Publish the raster budget the engine's memory ledger enforces. Once per
/// document, from the open flow: the numbers are the policy's, and nothing
/// about them depends on which book is open.
pub(crate) fn publish_budget() {
    engine::configure_motion(&MotionBudget {
        max_bytes: POLICY.memory.max_bytes as f64,
        preview_bytes: POLICY.memory.preview_bytes as f64,
        max_preview_pages: POLICY.memory.max_preview_pages as u32,
        preview_scale: POLICY.rendering.preview_scale,
    });
}

/// Wire one strip's virtualizer to the engine's scheduler.
///
/// An effect on the plan, and a comparison before the call: a scroll frame
/// that moved the reader within the same page produces the same payload, and
/// publishing it again would only re-score a queue against motion that has not
/// changed. What does re-publish is the settle — which is the frame that
/// matters most, because it is the one that promotes the preview ring.
pub(crate) fn use_motion_bridge(virtualizer: &Virtualizer) {
    let plan = virtualizer.render_plan();
    let published = StoredValue::new_local(None::<ScrollMotion>);
    Effect::new(move |_| {
        let motion = wire_motion(plan.get());
        if published.get_value() == Some(motion) {
            return;
        }
        published.set_value(Some(motion));
        engine::set_scroll_motion(&motion);
    });
}

#[cfg(test)]
mod tests {
    use super::*;
    use virtual_list_leptos::RenderQuality;

    fn window(first: usize, last: usize) -> Window {
        Window { first, last }
    }

    #[test]
    fn an_empty_document_publishes_no_motion() {
        // No mount window means no pages, and a prediction of "page 1" for a
        // document that has none would be a window the engine believes.
        let plan = RenderPlan {
            phase: ScrollPhase::Fling,
            direction: 1,
            predicted_index: 41,
            ..RenderPlan::default()
        };
        assert_eq!(wire_motion(plan), ScrollMotion::idle());
        assert_eq!(wire_motion(plan).phase, MotionPhase::Idle);
    }

    #[test]
    fn item_indices_cross_as_page_numbers() {
        let plan = RenderPlan {
            phase: ScrollPhase::Normal,
            direction: 1,
            predicted_index: 9,
            full: Some(window(8, 10)),
            preview: Some(window(5, 13)),
            mount: Some(window(2, 18)),
            delay_ms: 8,
            workers: 2,
            ..RenderPlan::default()
        };
        let motion = wire_motion(plan);
        // 0-based items become 1-based pages, and the window keeps its
        // inclusive ends — the off-by-one that would put the reader's own page
        // outside the window it is the centre of.
        assert_eq!(motion.predicted_page, 10);
        assert_eq!(motion.full, Some((9, 11)));
        assert_eq!(motion.preview, Some((6, 14)));
        assert_eq!(motion.delay_ms, 8);
        assert_eq!(motion.workers, 2);
        assert_eq!(motion.direction, 1);
    }

    #[test]
    fn every_phase_has_a_wire_spelling() {
        for (phase, want) in [
            (ScrollPhase::Idle, MotionPhase::Idle),
            (ScrollPhase::Slow, MotionPhase::Slow),
            (ScrollPhase::Normal, MotionPhase::Normal),
            (ScrollPhase::Fast, MotionPhase::Fast),
            (ScrollPhase::Fling, MotionPhase::Fling),
        ] {
            assert_eq!(wire_phase(phase), want);
        }
    }

    #[test]
    fn the_budget_is_the_policy_the_strips_run() {
        // The engine's ceiling and the virtualizer's tiers are one policy read
        // twice; if they ever disagree, the side that is wrong is the one that
        // stops matching this test.
        assert_eq!(POLICY.memory.max_bytes, 96 * 1024 * 1024);
        assert!(POLICY.memory.preview_bytes < POLICY.memory.max_bytes);
        assert!(POLICY.rendering.preview_scale > 0.0);
        assert!(POLICY.rendering.preview_screens > POLICY.rendering.full_screens);
        // The mount ceiling the strips are built with has to admit the preview
        // ring, or the ring is a policy for pages that are never mounted.
        assert!(POLICY.memory.max_preview_pages <= reader_core::view::MOUNTED_PAGES);
    }

    #[test]
    fn the_plan_the_bridge_publishes_tiers_its_own_pages() {
        // The end-to-end claim, on the pure half of it: a plan for a reader
        // moving down gives the page ahead of them a better tier than the one
        // behind, and the page they are predicted to land on is full quality
        // even when it sits past the full tier.
        let plan = RenderPlan {
            phase: ScrollPhase::Fast,
            direction: 1,
            predicted_index: 12,
            visible: Some(window(9, 10)),
            full: Some(window(9, 11)),
            preview: Some(window(8, 14)),
            mount: Some(window(4, 18)),
            ..RenderPlan::default()
        };
        assert_eq!(plan.quality(9), RenderQuality::Full);
        assert_eq!(plan.quality(11), RenderQuality::Full);
        assert_eq!(plan.quality(13), RenderQuality::Preview);
        assert_eq!(plan.quality(5), RenderQuality::Placeholder);
        assert_eq!(wire_motion(plan).predicted_page, 13);
    }
}
